lib/microcbor-derive/src/lib.rs

```diff
@@ -0,0 +1,1030 @@
+            input.attrs,
+            input.vis,
+            input.ident,
+            input.generics,
+            data.clone(),
+        )
+        .map(|tokens| tokens.to_token_stream().into()),
+        syn::Data::Struct(data) => gen_encode_struct_impl(
+            input.attrs,
+            input.vis,
+            input.ident,
+            input.generics,
+            data.clone(),
+        )
+        .map(|tokens| tokens.to_token_stream().into()),
+        _ => Err(syn::Error::new_spanned(
+            input,
+            "`StaticCborLen` can only be derived for `struct` and `enum` \
+             types",
+        )),
+    }
+}
+
+const HELPER_ATTR: &str = "cbor";
+const RENAME_ATTR: &str = "rename";
+const VARIANT_ID_ATTR: &str = "variant_id";
+
+fn gen_enum_encode_impl(
+    attrs: Vec<Attribute>,
+    _vis: Visibility,
+    ident: Ident,
+    generics: Generics,
+    data: DataEnum,
+) -> Result<impl ToTokens, syn::Error> {
+    // TODO(eliza): support top-level attribute for using the enum's repr
+    // instead of its name
+    let EnumDefAttrs {
+        variant_id_field_name,
+    } = EnumDefAttrs::parse(&attrs)?;
+    let mut variant_patterns = Vec::new();
+    let mut variant_lens = Vec::new();
+    let mut all_where_bounds = Vec::new();
+
+    for variant in data.variants {
+        let EnumVariantAttrs { rename } =
+            EnumVariantAttrs::parse(&variant.attrs)?;
+        let name = rename.unwrap_or_else(|| {
+            LitStr::new(&variant.ident.to_string(), variant.ident.span())
+        });
+
+        let variant_name = &variant.ident;
+        match variant.fields {
+            syn::Fields::Unit => match variant_id_field_name {
+                None => {
+                    variant_patterns.push(quote! {
+                        #ident::#variant_name => {
+                            __microcbor_renamed_encoder.str(#name)?;
+                        }
+                    });
+                    variant_lens.push(quote! {
+                        if ::microcbor::str_cbor_len(#name) > max {
+                            max = ::microcbor::str_cbor_len(#name);
+                        }
+                    });
+                }
+                Some(ref field_name) => {
+                    variant_patterns.push(quote! {
+                        #ident::#variant_name => {
+                            __microcbor_renamed_encoder
+                                .map(1)?
+                                .str(#field_name)?
+                                .str(#name)?;
+                        }
+                    });
+                    variant_lens.push(quote! {
+                        #[allow(non_snake_case)]
+                        let #variant_name = {
+                            // this will encode exactly 1 field, so we use the
+                            // length-prefixed repr to save a byte.
+                            let mut len = ::microcbor::u64_cbor_len(1);
```

Comment:
I would just write this as `1`, because it's a major type map + length 1, which is always encoded into a single byte (we're never actually encoding a `u64`).